File: 56-RADStudio10.3/AndroidAPILevelUpgrade_Demos/Object Pascal/Multi-Device Samples/Device Sensors and Services/Address Book/BirthdayReminder/MainFrm.pas
//---------------------------------------------------------------------------
// Copyright (c) 2016 Embarcadero Technologies, Inc. All rights reserved.  
//
// This software is the copyrighted property of Embarcadero Technologies, Inc. 
// ("Embarcadero") and its licensors. You may only use this software if you 
// are an authorized licensee of Delphi, C++Builder or RAD Studio 
// (the "Embarcadero Products").  This software is subject to Embarcadero's 
// standard software license and support agreement that accompanied your 
// purchase of the Embarcadero Products and is considered a Redistributable, 
// as such term is defined thereunder. Your use of this software constitutes 
// your acknowledgement of your agreement to the foregoing software license 
// and support agreement. 
//---------------------------------------------------------------------------
unit MainFrm;

interface

uses
  System.SysUtils, System.Types, System.UITypes, System.Classes, System.Variants, System.Permissions,
  FMX.Types, FMX.Controls, FMX.Forms, FMX.Graphics, FMX.Dialogs, FMX.Controls.Presentation, FMX.StdCtrls,
  System.Notification, FMX.ListView.Types, FMX.ListView.Appearances, FMX.ListView.Adapters.Base, FMX.ListView,
  FMX.AddressBook, FMX.TabControl, FMX.Calendar, FMX.DateTimeCtrls,
  FMX.ListBox, FMX.Layouts, System.ImageList, FMX.ImgList, FMX.Surfaces, FMX.Ani, FMX.Objects,
  System.Generics.Collections, System.Generics.Defaults, FMX.AddressBook.Types;

resourcestring
  NotificationMessage = 'Don''t forget to congratulate %s';

type
  TFetchContactThread = class;

  TFormMain = class(TForm)
    NotificationCenter1: TNotificationCenter;
    ListView1: TListView;
    ToolBar1: TToolBar;
    LabelTitle: TLabel;
    SpeedButton1: TSpeedButton;
    ImageList1: TImageList;
    AddressBook: TAddressBook;
    ProgressPanel: TRectangle;
    ProgressBar: TProgressBar;
    LabelProgress: TLabel;
    ProgressPanelAnimation: TFloatAnimation;
    procedure ListView1PullRefresh(Sender: TObject);
    procedure SpeedButton1Click(Sender: TObject);
    procedure AddressBookPermissionRequest(ASender: TObject; const AMessage: string; const AAccessGranted: Boolean);
    procedure FormShow(Sender: TObject);
    procedure AddressBookExternalChange(ASender: TObject);
    procedure FormDestroy(Sender: TObject);
  private
    FContacts: TAddressBookContacts;
    FThread: TFetchContactThread;
    FPermissionReadContacts: string;
    procedure FillContactsList;
    function CreateRoundPhoto(const ASource: TBitmapSurface): TBitmap;
    procedure PostNotification(const ADisplayName: string; const ABirthday: TDateTime; const ARemainderDays: Integer);
    function DefineRemainedDays(const ABirthday: TDate): Integer;
    function ChangeYear(const ADateTime: TDateTime; const ANewYear: Word): TDateTime;
    { TFetchContactThread events handlers }
    procedure ContactLoadingBegin(Sender: TObject);
    procedure ContactLoaded(const ATotalCount:Integer; const ANumber: Integer; const ABirthday: TDateTime; const ADisplayName: string;
      const APhoto: TBitmapSurface);
    procedure ContactLoadingEnd(Sender: TObject);
    procedure DisplayRationale(Sender: TObject; const APermissions: TArray<string>; const APostRationaleProc: TProc);
  public
    procedure SortContacts;
  end;

  TFetchContactThread = class(TThread)
  public type
    TOnContactLoaded = procedure (const ATotalCount: Integer; const ANumber: Integer; const ABirthday: TDateTime;
      const ADisplayName: string; const APhoto: TBitmapSurface) of object;
  private
    [Weak] FAllContacts: TAddressBookContacts;
    FOnStart: TNotifyEvent;
    FOnContactLoaded: TOnContactLoaded;
  protected
    procedure Execute; override;
    procedure DoStart;
    procedure DoContactLoaded(const ANumber: Integer; const ABirthday: TDateTime; const ADisplayName: string;
      const APhoto: TBitmapSurface);
  public
    constructor Create(const AContacts: TAddressBookContacts);
    property OnContactLoaded: TOnContactLoaded read FOnContactLoaded write FOnContactLoaded;
    property OnStart: TNotifyEvent read FOnStart write FOnStart;
  end;

var
  FormMain: TFormMain;

implementation

uses
{$IFDEF ANDROID}
  Androidapi.Helpers,
  Androidapi.JNI.JavaTypes,
  Androidapi.JNI.Os,
{$ENDIF}
  System.DateUtils, System.Math, FMX.DialogService;

{$R *.fmx}

const
  Months: array [1..12] of string = ('January', 'February', 'March', 'April', 'May', 'June', 'July', 'August',
    'September', 'October', 'November', 'December');

{ TFormMain }

procedure TFormMain.AddressBookExternalChange(ASender: TObject);
begin
  AddressBook.RevertCurrentChangesAndUpdate;
end;

procedure TFormMain.AddressBookPermissionRequest(ASender: TObject; const AMessage: string;
  const AAccessGranted: Boolean);
begin
  if AAccessGranted then
    FillContactsList
  else
  begin
    ProgressPanelAnimation.Enabled := True;
    ProgressPanelAnimation.Start;
    ProgressBar.Visible := False;
    LabelProgress.Text := 'User not allowed to read contacts. '+ AMessage;
  end;
end;

function TFormMain.ChangeYear(const ADateTime: TDateTime; const ANewYear: Word): TDateTime;
var
  Year: Word;
  Month: Word;
  Days: Word;
begin
  DecodeDate(ADateTime, Year, Month, Days);
  if Month = MonthFebruary then
    Days := Min(Days, DaysInAMonth(ANewYear, MonthFebruary));
  Result := EncodeDate(ANewYear, Month, Days);
  ReplaceTime(Result, ADateTime);
end;

procedure TFormMain.ContactLoaded(const ATotalCount: Integer; const ANumber: Integer; const ABirthday: TDateTime; const ADisplayName: string;
  const APhoto: TBitmapSurface);
var
  Item: TListViewItem;
  RemainderDays: Integer;
begin
  if not IsNan(ABirthday) then
  begin
    RemainderDays := DefineRemainedDays(ABirthday);
    Item := ListView1.Items.Add;
    Item.Detail := FormatDateTime('dd mmmm', ABirthday);
    Item.Text := ADisplayName;
    Item.Tag := RemainderDays;
    Item.ButtonText := Format('%d days', [RemainderDays]);
    if APhoto = nil then
      Item.ImageIndex := 0
    else
      Item.Bitmap := CreateRoundPhoto(APhoto);

    PostNotification(ADisplayName, ABirthday, RemainderDays);
    SortContacts;
  end;
  ProgressBar.Value := ANumber + 1;  // Number start from 0, so we need to add 1
  LabelProgress.Text := Format('Loaded %d from %d contacts', [ANumber, ATotalCount]);
end;

procedure TFormMain.ContactLoadingBegin(Sender: TObject);
begin
  ListView1.Items.Clear;
  ListView1.PullRefreshWait := True;
  NotificationCenter1.CancelAll;
  ProgressPanelAnimation.Inverse := False;
  ProgressPanelAnimation.Enabled := True;
  ProgressPanelAnimation.Start;
  ProgressBar.Value := 0;
  ProgressBar.Max := FContacts.Count;
end;

procedure TFormMain.ContactLoadingEnd(Sender: TObject);
begin
  ListView1.PullRefreshWait := False;
  ProgressPanelAnimation.Inverse := True;
  ProgressPanelAnimation.Start;
end;

function TFormMain.CreateRoundPhoto(const ASource: TBitmapSurface): TBitmap;
var
  Size: Integer;
begin
  Result := TBitmap.Create;
  Size := Min(ASource.Height, ASource.Width);
  Result.SetSize(Size, Size);
  Result.Clear(TAlphaColorRec.Null);
  if Result.Canvas.BeginScene then
    try
      Result.Canvas.Fill.Bitmap.Bitmap.Assign(ASource);
      Result.Canvas.Fill.Kind := TBrushKind.Bitmap;
      Result.Canvas.FillEllipse(TRectF.Create(0, 0, Size, Size), 1);
    finally
      Result.Canvas.EndScene;
    end;
end;

function TFormMain.DefineRemainedDays(const ABirthday: TDate): Integer;
var
  EventDate: TDateTime;
begin
  EventDate := ChangeYear(ABirthday, CurrentYear);
  Result := DaysBetween(Date, EventDate);
  if (Result > 0) and (EventDate < Now) then
    Result := DaysInYear(EventDate) - Result - 1;
end;

// Optional rationale display routine to display permission requirement rationale to the user
procedure TFormMain.DisplayRationale(Sender: TObject; const APermissions: TArray<string>; const APostRationaleProc: TProc);
var
  I: Integer;
  RationaleMsg: string;
begin
  for I := 0 to Pred(Length(APermissions)) do
  begin
    if APermissions[I] = FPermissionReadContacts then
      RationaleMsg := RationaleMsg + 'The app needs to look in the address book';
    if I <> Pred(Length(APermissions)) then
      RationaleMsg := RationaleMsg + SLineBreak + SLineBreak;
  end;

  // Show an explanation to the user *asynchronously* - don't block this thread waiting for the user's response!
  // After the user sees the explanation, invoke the post-rationale routine to request the permissions
  TDialogService.ShowMessage(RationaleMsg,
    procedure(const AResult: TModalResult)
    begin
      APostRationaleProc
    end)
end;

procedure TFormMain.FillContactsList;
begin
  FreeAndNil(FContacts);
  FContacts := TAddressBookContacts.Create(True);
  AddressBook.AllContacts(FContacts);

  if (FThread <> nil) and not FThread.Finished then
  begin
    FThread.Terminate;
    FThread.WaitFor;
  end;
  FreeAndNil(FThread);

  FThread := TFetchContactThread.Create(FContacts);
  FThread.OnContactLoaded := ContactLoaded;
  FThread.OnStart := ContactLoadingBegin;
  FThread.OnTerminate := ContactLoadingEnd;
  FThread.Start;
end;

procedure TFormMain.FormDestroy(Sender: TObject);
begin
  if (FThread <> nil) and not FThread.Finished then
  begin
    FThread.OnContactLoaded := nil;
    FThread.Terminate;
  end;
end;

procedure TFormMain.FormShow(Sender: TObject);
begin
{$IFDEF ANDROID}
  FPermissionReadContacts := JStringToString(TJManifest_permission.JavaClass.READ_CONTACTS);
{$ENDIF}
  AddressBook.RequestPermission(DisplayRationale);
end;

procedure TFormMain.ListView1PullRefresh(Sender: TObject);
begin
  ListView1.PullRefreshWait := True;
  FillContactsList;
end;

procedure TFormMain.PostNotification(const ADisplayName: string; const ABirthday: TDateTime;
  const ARemainderDays: Integer);
var
  Notification: TNotification;
  EventDate: TDateTime;
begin
  Notification := NotificationCenter1.CreateNotification;
  try
    EventDate := ABirthday;
    EventDate := RecodeTime(EventDate, 11, 0, 0, 0);
    EventDate := ChangeYear(EventDate, CurrentYear);
    if EventDate < Now then
      EventDate := ChangeYear(EventDate, CurrentYear + 1);
    Notification.RepeatInterval := TRepeatInterval.Year;
    Notification.FireDate := EventDate;
    Notification.AlertBody := Format(NotificationMessage, [ADisplayName]);
    Notification.EnableSound := True;
    NotificationCenter1.ScheduleNotification(Notification);
  finally
    Notification.Free;
  end;
end;

procedure TFormMain.SortContacts;
begin
  ListView1.BeginUpdate;
  try
    ListView1.Items.Sort(TComparer<TListViewItem>.Construct(
      function(const Left, Right: TListViewItem): Integer
      begin
        Result := CompareValue(Left.Tag, Right.Tag);
      end));
  finally
    ListView1.EndUpdate;
  end;
end;

procedure TFormMain.SpeedButton1Click(Sender: TObject);
begin
  ListView1.SearchVisible := not ListView1.SearchVisible;
end;

{ TFetchContactThread }

constructor TFetchContactThread.Create(const AContacts: TAddressBookContacts);
begin
  inherited Create(True);
  FAllContacts := AContacts;
end;

procedure TFetchContactThread.DoContactLoaded(const ANumber: Integer; const ABirthday: TDateTime; const ADisplayName: string;
  const APhoto: TBitmapSurface);
begin
  if Assigned(FOnContactLoaded) then
    FOnContactLoaded(FAllContacts.Count, ANumber, ABirthday, ADisplayName, APhoto);
end;

procedure TFetchContactThread.DoStart;
begin
  if Assigned(FOnStart) then
    FOnStart(Self);
end;

procedure TFetchContactThread.Execute;
var
  Contact: TAddressBookContact;
  Index: Integer;
  DisplayName: string;
  Photo: TBitmapSurface;
  Birthday: TDateTime;
begin
  Synchronize(DoStart);
  Index := 0;
  for Contact in FAllContacts do
  begin
    if Terminated then
      Exit;

    DisplayName := Contact.DisplayName;
    Photo := Contact.PhotoThumbnail;
    Birthday := Contact.Birthday;
    try
      Synchronize(procedure
        begin
          DoContactLoaded(Index, Birthday, DisplayName, Photo);
        end);
    finally
      if Photo <> nil then
        Photo.Free;
    end;
    Inc(Index);
  end;
end;

end.

File: 56-RADStudio10.3/AndroidAPILevelUpgrade_Demos/CPP/Multi-Device Samples/Device Sensors and Services/Address Book/BirthdayReminder/MainFrm.cpp
// ---------------------------------------------------------------------------
// Copyright (c) 2016 Embarcadero Technologies, Inc. All rights reserved.
//
// This software is the copyrighted property of Embarcadero Technologies, Inc.
// ("Embarcadero") and its licensors. You may only use this software if you
// are an authorized licensee of Delphi, C++Builder or RAD Studio
// (the "Embarcadero Products").  This software is subject to Embarcadero's
// standard software license and support agreement that accompanied your
// purchase of the Embarcadero Products and is considered a Redistributable,
// as such term is defined thereunder. Your use of this software constitutes
// your acknowledgement of your agreement to the foregoing software license
// and support agreement.
// ---------------------------------------------------------------------------
// ---------------------------------------------------------------------------

#include <fmx.h>
#include <System.Permissions.hpp>
#ifdef __ANDROID__
	#include <Androidapi.Helpers.hpp>
	#include <Androidapi.JNI.Os.hpp>
#endif
#pragma hdrstop

#include "MainFrm.h"
#include <System.UITypes.hpp>
#include <System.DateUtils.hpp>
#include <FMX.DialogService.hpp>
// ---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.fmx"
TFormMain *FormMain;

// ---------------------------------------------------------------------------
__fastcall TFormMain::TFormMain(TComponent* Owner) : TForm(Owner) {
}

// ---------------------------------------------------------------------------
void __fastcall TFormMain::AddressBookExternalChange(TObject *ASender) {
	AddressBook->RevertCurrentChangesAndUpdate();
}

// ---------------------------------------------------------------------------
void __fastcall TFormMain::FormShow(TObject *Sender) {
#ifdef __ANDROID__
    FPermissionReadContacts = JStringToString(TJManifest_permission::JavaClass->READ_CONTACTS);
#endif
    AddressBook->RequestPermission(DisplayRationale);
}

// ---------------------------------------------------------------------------

// Optional rationale display routine to display permission requirement rationale to the user
void __fastcall TFormMain::DisplayRationale(TObject* Sender, const TStringDynArray APermissions, const _di_TProc APostRationaleProc) {
	String RationaleMsg;
	for (int i = 0; i < APermissions.Length; i++)
	{
        if (APermissions[i] == FPermissionReadContacts)
            RationaleMsg = RationaleMsg + "The app needs to look in the address book";
		RationaleMsg = RationaleMsg + APermissions[i];
		if (i != APermissions.Length-1)
			RationaleMsg = RationaleMsg + sLineBreak + sLineBreak;
	}

    // Show an explanation to the user *asynchronously* - don't block this thread waiting for the user's response!
    // After the user sees the explanation, invoke the post-rationale routine to request the permissions
    TDialogService::ShowMessage(RationaleMsg,
        [APostRationaleProc](TModalResult AKey)
        {
            APostRationaleProc->Invoke();
        });
}
// ---------------------------------------------------------------------------

void __fastcall TFormMain::ListView1PullRefresh(TObject *Sender) {
	ListView1->PullRefreshWait = True;
	this->FillContactsList();
}

// ---------------------------------------------------------------------------
void TFormMain::FillContactsList() {
	if ((thread != NULL) && (!thread->Finished)) {
		thread->Terminate();
		thread->WaitFor();
	}
	if (thread != NULL) {
		thread->Free();
	}
	if (contacts != NULL) {
		contacts->Free();
	}
	contacts = new TAddressBookContacts();
	AddressBook->AllContacts(contacts);

	thread = new TFetchContactThread(contacts);
	thread->OnContactLoaded = ContactLoaded;
	thread->OnStart = ContactLoadingBegin;
	thread->OnTerminate = ContactLoadingEnd;
	thread->Start();
}

// ---------------------------------------------------------------------------
TBitmap* TFormMain::CreateRoundPhoto(TBitmapSurface* source) {
	int size = Min(source->Width, source->Height);
	TBitmap* bitmap = new TBitmap();
	bitmap->SetSize(size, size);
	bitmap->Clear(0);
	if (bitmap->Canvas->BeginScene()) {
		try {
			bitmap->Canvas->Fill->Bitmap->Bitmap->Assign(source);
			bitmap->Canvas->Fill->Kind = TBrushKind::Bitmap;
			bitmap->Canvas->FillEllipse(TRectF(0, 0, size, size), 1);
		}
		__finally {
			bitmap->Canvas->EndScene();
		}
	}
	return bitmap;
}

// ---------------------------------------------------------------------------
void TFormMain::PostNotification(const UnicodeString displayName, const TDateTime birthday, const int remainderDays) {

	TNotification* notification = NotificationCenter1->CreateNotification();
	try {
		TDateTime eventDate = birthday;
		eventDate = RecodeTime(eventDate, 11, 0, 0, 0);
		eventDate = ChangeYear(eventDate, CurrentYear());
		if (eventDate < Now())
			eventDate = ChangeYear(eventDate, CurrentYear() + 1);
		notification->RepeatInterval = TRepeatInterval::Year;
		notification->FireDate = eventDate;
		notification->AlertBody = "Don't forget to congratulate " + displayName;
		notification->EnableSound = true;
		NotificationCenter1->ScheduleNotification(notification);
	}
	__finally {
		notification->Free();
	}
}

// ---------------------------------------------------------------------------
int TFormMain::DefineRemainedDays(const TDate birthday) {
	TDateTime eventDate = ChangeYear(birthday, CurrentYear());
	int days = DaysBetween(Date(), eventDate);
	if ((days > 0) && (eventDate < Now())) {
		days = DaysInYear(eventDate) - days - 1;
	}
	return days;
}

// ---------------------------------------------------------------------------
TDateTime TFormMain::ChangeYear(const TDateTime date, const unsigned short newYear) {
	unsigned short year, month, days;
	DecodeDate(date, year, month, days);
	if (month == MonthFebruary) {
		days = Min(days, DaysInAMonth(newYear, MonthFebruary));
	}
	TDateTime newDate = EncodeDate(newYear, month, days);
	ReplaceTime(newDate,date);
	return newDate;
}

// ---------------------------------------------------------------------------
void __fastcall TFormMain::ContactLoadingBegin(TObject *Sender) {
	ListView1->Items->Clear();
	ListView1->PullRefreshWait = true;
	NotificationCenter1->CancelAll();
	ProgressPanelAnimation->Inverse = false;
	ProgressPanelAnimation->Enabled = true;
	ProgressPanelAnimation->Start();
	ProgressBar->Value = 0;
	ProgressBar->Max = contacts->Count;
}

// ---------------------------------------------------------------------------
void __fastcall TFormMain::ContactLoaded(const int totalCount, const int number, const TDateTime birthday,
	const UnicodeString displayName, TBitmapSurface* photo) {
	TListViewItem* item;
	int remainderDays;

	if (!IsNan(birthday.Val)) {
		remainderDays = DefineRemainedDays(birthday);
		item = ListView1->Items->Add();
		item->Detail = FormatDateTime("dd mmmm", birthday);
		item->Text = displayName;
		item->Tag = remainderDays;
		TVarRec vr[] = {remainderDays};
		item->ButtonText = Format("%d days", vr, 1);
		if (photo == NULL) {
			item->ImageIndex = 0;
		}
		else {
			item->Bitmap = CreateRoundPhoto(photo);
		}
		PostNotification(displayName, birthday, remainderDays);
	}
	ProgressBar->Value = number + 1; // Number start from 0, so we need to add 1
	TVarRec vr[] = {number, totalCount};
	LabelProgress->Text = Format("Loaded %d from %d contacts", vr, 2);
}

// ---------------------------------------------------------------------------
void __fastcall TFormMain::ContactLoadingEnd(TObject* Sender) {
	ListView1->PullRefreshWait = false;
	ProgressPanelAnimation->Inverse = true;
	ProgressPanelAnimation->Start();
}

// ---------------------------------------------------------------------------
void __fastcall TFormMain::SpeedButton1Click(TObject *Sender) {
	ListView1->SearchVisible = !ListView1->SearchVisible;
}

// ---------------------------------------------------------------------------
void __fastcall TFormMain::AddressBookPermissionRequest(TObject *ASender, const UnicodeString AMessage,
	const bool AAccessGranted) {
	if (AAccessGranted) {
		FillContactsList();
	}
	else {
		ProgressPanelAnimation->Enabled = true;
		ProgressPanelAnimation->Start();
		ProgressBar->Visible = false;
		LabelProgress->Text = "User not allowed to read contacts. " + AMessage;
	}
}
// ---------------------------------------------------------------------------
void __fastcall TFormMain::FormDestroy(TObject *Sender) {
    if ((thread != NULL) && (!thread->Finished)) {
        thread->OnContactLoaded = (TOnContactLoaded)NULL;
        thread->Terminate();
    }
}
//---------------------------------------------------------------------------

